driver-core/src/main/com/mongodb/internal/capi/MongoCryptHelper.java

```diff
@@ -40,13 +43,36 @@ public static MongoCryptOptions createMongoCryptOptions(final Map<String, Map<St
                                                             final Map<String, BsonDocument> namespaceToLocalSchemaDocumentMap) {
         MongoCryptOptions.Builder mongoCryptOptionsBuilder = MongoCryptOptions.builder();
 
+        BsonDocument bsonKmsProviders = getKmsProvidersAsBsonDocument(kmsProviders);
+        mongoCryptOptionsBuilder.kmsProviderOptions(bsonKmsProviders);
+        mongoCryptOptionsBuilder.localSchemaMap(namespaceToLocalSchemaDocumentMap);
+        mongoCryptOptionsBuilder.needsKmsCredentialsStateEnabled(true);
+        return mongoCryptOptionsBuilder.build();
+    }
+
+    public static BsonDocument fetchCredentials(final Map<String, Map<String, Object>> kmsProviders) {
+        BsonDocument kmsProvidersDocument = MongoCryptHelper.getKmsProvidersAsBsonDocument(kmsProviders);
+        if (kmsProvidersDocument.containsKey("aws") && kmsProvidersDocument.get("aws").asDocument().isEmpty()) {
+            AwsCredential awsCredential = AwsCredentialHelper.obtainFromEnvironment();
```

Comment:
`AwsCredentialHelper.obtainFromEnvironment` may call `AwsCredentialHelper.obtainFromEc2OrEcsResponse`, which is a blocking call because it does IO. As a result, the method `com.mongodb.reactivestreams.client.internal.crypt.Crypt.fetchCredentials` is blocking even though it is also asynchronous. Consequently, `com.mongodb.reactivestreams.client.internal.crypt.Crypt.encrypt`/`decrypt` are now blocking, though previously they were both asynchronous and non-blocking. This, in turn, correspondingly affects the properties of the driver reactive API.

I don't have good enough understanding to say whether it's a problem or not (maybe our reactive API blocks in some situations even without this change), but it is worth thinking about or discussing.

Reply:
It's a problem that I'm aware of.  We already do the same thing in a couple of `SaslClient` implementations, including the one in AwsAuthenticator which uses the same helper method.